### ISSUE TITLE
cups-browsed incorrectly remove printers.

### ISSUE THREAD

**Issue by dheeraj135** [OpenPrinting/cups-filters#136]
I turn on my laptop's wifi.
I create a virtual printer with the command: ```ippserver -p 8010 t1 ```.
It creates two printers:
```
ipp printer at ipp://127.0.0.1:8010/ipp/print
ipps printer at ipps://127.0.0.1:8010/ipp/print
```
The printer is detected by cups-browsed and queue is automatically created for the ipps protocol. It doesn't create queue for the ipp protocol. The ipp printer is available under discovered network printers on the cups web interface.

I turn off my laptop's wifi and wait for cups-browsed to detect this network change(one loop cycle). The printer queue for ipps protocol is deleted. Even after the queue is deleted, The ipps printer is **not** available under discovered network printers on cups web interface.
Relevant lines from cups-browsed logs:

```
Fri Jul 26 01:08:06 2019 browse_callback() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 Avahi Browser: REMOVE: service 't1' of type '_ipps._tcp' in domain 'local' on interface 'enp0s20f0u1'
Fri Jul 26 01:08:06 2019 Printer t1 (Host: 127.0.0.1, Port: 8010, URI: ipps://127.0.0.1:8010/ipp/print) disappeared and no slave available (or it is a slave of another printer), removing entry.
Fri Jul 26 01:08:06 2019 DNS-SD IDs: Service name: "t1", Service type: "_ipps._tcp", Domain: "local"
Fri Jul 26 01:08:06 2019 checking queues in 0s
Fri Jul 26 01:08:06 2019 browse_callback() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 Avahi Browser: REMOVE: service 't1' of type '_ipp._tcp' in domain 'local' on interface 'enp0s20f0u1'
Fri Jul 26 01:08:06 2019 nm_properties_changed() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 nm_properties_changed() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 NetworkManager ActiveConnections changed
Fri Jul 26 01:08:06 2019 nm_properties_changed() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 nm_properties_changed() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 browse_callback() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 Avahi Browser: REMOVE: service 't1' of type '_ipps._tcp' in domain 'local' on interface 'enp0s20f0u1'
Fri Jul 26 01:08:06 2019 browse_callback() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 Avahi Browser: REMOVE: service 't1' of type '_ipp._tcp' in domain 'local' on interface 'enp0s20f0u1'
Fri Jul 26 01:08:06 2019 nm_properties_changed() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 NetworkManager ActiveConnections changed
Fri Jul 26 01:08:06 2019 update_cups_queues() in THREAD 140606784412480
Fri Jul 26 01:08:06 2019 Processing printer list ...
Fri Jul 26 01:08:06 2019 === Remote printer overview ===
Fri Jul 26 01:08:06 2019 Printer ipps://127.0.0.1:8010/ipp/print (127.0.0.1, 127.0.0.1): Local queue t1@127.0.0.1, IPP Printer, Slave of None (Disappeared)
Fri Jul 26 01:08:06 2019 Printer ipps://127.0.0.1:8010/ipp/print (127.0.0.1, 127.0.0.1): Local queue t1, IPP Printer, Slave of None (Disappeared)

```

**Comment by tillkamppeter** [comment]
cups-browsed does everything to avoid creating two queues for only one physical printer. Here we have a printer supporting both ipp: and ipps: protocols. cups-browsed recognizes that these two URIs are only one physical printer and only creates a queue for the "better" version, which is ipps:. When the printer disappears, it only removes the queue it has created, the one for ipps:. So in terms of real (no virtual, temporary,local, ...) print queues (queues which "`lpstat -v`" would list) cups-browsed does the right thing and there should only be the ipps: queue as real queue.
So while the printer is available, "`lpstat -v`" should only list the ipps: queue whereas "`lpstat -e`" lists  both.
If you stoip cups-browsed, the physical ipps: queue will get deleted ("`lpstat -v`" does not list it any more) but it should continue as a virtual CUPS queue listed by "`lpstat -e`", as long as CUPS lists both queue versions for the same physical printer and not also only one, but this would be a CUPS bug/feature.
But why does your printer disappear when turning off your laptop's WiFi, the printer is on localhost.
Could you check with avahi-discover whether your printer is correctly broadcasted?
Does your Avahi correctly support localhost? See https://github.com/OpenPrinting/ippusbxd/blob/master/readme.md for info about the problem. Current Ubuntu has the patch.

**Comment by dheeraj135** [comment]
The printer is correctly broadcasted.

Result of avahi-discover when wifi is on:
```
Found service 't1' of type '_http._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_printer._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_http._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_printer._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_http._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_printer._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_http._tcp' in domain 'local' on 3.1.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 3.1.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 3.1.
Found service 't1' of type '_printer._tcp' in domain 'local' on 3.1.
Found service 't1' of type '_http._tcp' in domain 'local' on 3.0.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 3.0.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 3.0.
Found service 't1' of type '_printer._tcp' in domain 'local' on 3.0.
Found service 't1' of type '_http._tcp' in domain 'local' on 1.0.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 1.0.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 1.0.
Found service 't1' of type '_printer._tcp' in domain 'local' on 1.0.
```

Result of avahi-discover when wifi is off:
```
Found service 't1' of type '_http._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_printer._tcp' in domain 'local' on 7.1.
Found service 't1' of type '_http._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_printer._tcp' in domain 'local' on 5.1.
Found service 't1' of type '_http._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_printer._tcp' in domain 'local' on 5.0.
Found service 't1' of type '_http._tcp' in domain 'local' on 1.0.
Found service 't1' of type '_ipps._tcp' in domain 'local' on 1.0.
Found service 't1' of type '_ipp._tcp' in domain 'local' on 1.0.
Found service 't1' of type '_printer._tcp' in domain 'local' on 1.0.
```
I am using ubuntu 19.04 and I installed avahi using apt. So, I am assuming avahi is already patched.

**Comment by tillkamppeter** [comment]
What is exactly going wrong here? That cups-browsed is removing its queue when turning off Wi-Fi (printer is on lo and cups-browsed should also work in lo-only environment)? CUPS not showing both ipp: and ipps: version of printer in "`lpstat -e`" (I actually do not know what CUPS should show here)?
Please provide a debug log of cups-browsed covering the point in time when the unexpected happens. If there is not already a `/var/log/cups/cups-browsed_log` file, make sure to have a line `DebugLogging file` in `/etc/cups/cups-browsed.conf` and restart cups-browsed.
Rename your log file to have `.txt` extension and attach it to this issue, do not compress it.

**Comment by dheeraj135** [comment]
cups-browsed removing its queue when turning off Wi-Fi is the only issue.
Log file: 
[cups-browsed.txt](https://github.com/OpenPrinting/cups-filters/files/3433399/cups-browsed.txt)
Events logged into the log file: cups-browsed being restarted, t1 is discovered. Then wifi is turned off and print queue for t1 is deleted.

**Comment by tillkamppeter** [comment]
Do you have an actual, physical printer?
Can you also post your file `/etc/cups/cups-browsed.conf`? Please rename it to have a `.txt` extension for attaching here.

**Comment by tillkamppeter** [comment]
Are you sure trhat you are running only one printer emulation in your local network? Or could there be a second instance of ippserver running on a machine named "dj-HP.loca"?

**Comment by dheeraj135** [comment]
> Do you have an actual, physical printer?
Yes, I have HP Officejet 6960. But I cannot recreate this issue on a physical printer because for this we will need the printer to be connected on more than one interfaces. In the case of ippserver, the printer is connected on 'lo', and wlo1 interfaces.
> Can you also post your file `/etc/cups/cups-browsed.conf`? Please rename it to have a `.txt` extension for attaching here.
[cups-browsed.txt](https://github.com/OpenPrinting/cups-filters/files/3437967/cups-browsed.txt)
> Are you sure trhat you are running only one printer emulation in your local network? Or could there be a second instance of ippserver running on a machine named "dj-HP.loca"?
Yes, I am sure. If I turn off the wifi and restart the ippserver, the printer is detected normally detected by the cups-browsed and print queue is created.
According to me, cups-browsed/avahi attaches each printer to a specific interface. If a printer is connected on multiple interfaces, it just picks one of them. If that interface is disconnected( in my case wlo1), cups-browsed/avahi deletes all printers connected using that interface. But we need to handle cases where a printer is connected on multiple interfaces.

**Comment by tillkamppeter** [comment]
cups-browsed creates one entry for each printer it discovers, also if it discovers one and the same printer through several different interfaces. If one interface goes away, the printer entry (and so the local queue) is removed, even if the printer is still available through another interface.
To cope with this, the following improvements/fixes on cups-browsed are needed:
- The printer entry needs a list of interfaces through which this printer got discovered. On each discovery of a printer which was already discovered, the new interface needs to get added to the list and on a removal event the interface needs to get removed, but the entry should only get removed when no interface is left.
- If a printer is discovered via the "lo" interface this should be preferred, setting the host name in the URI to "localhost" or "127.0.0.1". This avoids that communication with a local service does detours through routers or APs (could perhaps also reveal private data to outside) and also "lo" is always there, from boot to shutdown, whereas other interfaces can get taken down or re-established mid-session.

**Comment by tillkamppeter** [comment]
Thank you very much for your bug report. This was an important problem, especially as we now will see emulated IPP printers on the local machine.

Fixed with commit 1a8cedd6fa.

I have solved this by registering each DNS-SD message of a printer appearing with network interface, IPP type, and network family (IPv4/IPv6). The printer entry will only get actually removed if a corresponding disappearing DNS-SD message for each of the registered messages shows up. This way only a few interfaces disappearing makes the queue remain if the printer is available through other interfaces.
